lib/Cro/Tools/Template/ReactReduxSPA.pm6

```diff
@@ -0,0 +1,253 @@
+        self.write-webpack-config($where.add('webpack.config.js'));
+        self.write-babelrc($where.add('.babelrc'));
+        nextsame;
+    }
+
+    method new-directories($where) {
+        my $static = $where.add('static');
+        |callsame,
+        static    => $static,
+        static-js => $static.add('js'),
+        frontend  => $where.add('frontend')
+    }
+
+    method write-static-index($file, $name) {
+        $file.spurt(self.static-index-contents($name));
+    }
+
+    method static-index-contents($name) {
+        q:c:to/HTML/;
+            <html>
+              <head>
+                <title>{$name}</title>
+              </head>
+              <body>
+                <h1>{$name}</h1>
+                <div id="app"></div>
+                <script src="js/bundle.js"></script>
+              </body>
+            </html>
+            HTML
+    }
+
+    method write-frontend-index($file, $id, %options) {
+        $file.spurt(self.frontend-index-contents($id, %options));
+    }
+
+    method frontend-index-contents($id, %options) {
+        my $property = $id ~ 'Text';
+        my $chunk    = $id.tclc ~ 'Text';
+        my $event    = 'onChange' ~ $chunk;
+        my $function = 'change'  ~ $id    ~ 'Text';
+        my $reducer  = $id ~ 'Reducer';
+        my $contents = q:to/CODE/;
+            import React from 'react';
+            import { render } from 'react-dom';
+            import { createStore, applyMiddleware } from 'redux';
+            import { Provider, connect } from 'react-redux';
+            import thunkMiddleware from 'redux-thunk';
+            CODE
+        $contents ~= q:to/CODE/ if %options<websocket>;
+            import WSAction from 'redux-websocket-action';
+            CODE
+        $contents ~= q:s:to/CODE/;
+            import * as Actions from './actions';
+            import { $reducer } from './reducer';
+
+            var $chunk = props => (
+                <div>
+                    <h2>Text Editing</h2>
+                    <div>
+                        <textarea rows="2" cols="100" maxLength="200"
+                            value={props.$property}
+                            onChange={e => props.\qq[$event](e.target.value)} />
+                    </div>
+                </div>
+            );
+
+            var App = props => (
+                <div>
+                    <$chunk $property={props.$property}
+                        $event={props.$event} />
+                </div>
+            );
+
+            function mapProps(state) {
+                return state;
+            }
+            function mapDispatch(dispatch) {
+                return {
+                    $event: text => dispatch(Actions.\qq[$function](text)),
+                };
+            }
+
+            let store = createStore($reducer, applyMiddleware(thunkMiddleware));
+
+            CODE
+        $contents ~= q:to/CODE/ if %options<websocket>;
+            [].forEach(endpoint => {
```

Comment:
This loop over an empty array is a no-op; forget to include something?

Reply:
Two answers:

1. Nope, it's (intentionally) a stub to be filled in by the user when they have defined websocket endpoints.
2. This could be filled in better even in the stub form by explicitly hooking up the "chat" websocket code on both ends, which I have not (yet) done.

Do you want me to do #2 before merging the PR, or is #1 satisfying to you already?

Reply:
I'd prefer #2 if it's OK with you, though if you'd much prefer to put it off until later, I can merge this as is.